examples/Processor.hs

```diff
@@ -77,29 +94,40 @@ data RW a = R Key                 (Value -> a)
 -- | A program is a free selective on top of the 'RW' base functor.
 type Program a = Select RW a
 
-instance Show a => Show (RW a) where
+instance Show (RW a) where
     show (R k          _) = "Read "  ++ show k
     show (W k (Pure v) _) = "Write " ++ show k ++ " " ++ show v
     show (W k _        _) = "Write " ++ show k
 
+logEntry :: MonadState State m => LogEntry Key Value -> m ()
+logEntry item = S.modify $ \s ->
+    s {log = item : log s }
+
 -- | Interpret the base functor in a 'MonadState'.
 toState :: MonadState State m => RW a -> m a
-toState (R k t) = t <$> case k of
-    Reg  r    -> (Map.! r   ) <$> S.gets registers
-    Cell addr -> (Map.! addr) <$> S.gets memory
-    Flag f    -> (Map.! f   ) <$> S.gets flags
-    PC        -> pc <$> S.get
-toState (W k p t) = case k of
-    Reg r     -> do v <- runSelect toState p
-                    let regs' s = Map.insert r v (registers s)
-                    S.state (\s -> (t v, s {registers = regs' s}))
-    Cell addr -> do v <- runSelect toState p
-                    let mem' s = Map.insert addr v (memory s)
-                    S.state (\s -> (t v, s {memory = mem' s}))
-    Flag f    -> do v <- runSelect toState p
-                    let flags' s = Map.insert f v (flags s)
-                    S.state (\s -> (t v, s {flags = flags' s}))
-    PC          -> error "toState: Can't write the Program Counter (PC)"
+toState = \case
+    (R k t) -> do
+        v <- case k of
+                   Reg  r    -> (Map.! r   ) <$> S.gets registers
+                   Cell addr -> (Map.! addr) <$> S.gets memory
+                   Flag f    -> (Map.! f   ) <$> S.gets flags
+                   PC        -> pc <$> S.get
+        logEntry (ReadEntry k v)
+        pure (t v)
+    (W k p t) -> do
+        v <- runSelect toState p
+        -- To log the value, we need to evaluate it first, thus the
+        -- writes gets logged in front of the reads. Note that this order is
+        -- different from what static analysis via 'getProgramEffects' shows
```

Comment:
I don't understand this comment. To me it looks like all reads/writes are recorded in the correct order: we first evaluate the program that computes the value to be written (`v <- runSelect toState p`), recording all associated reads/writes, and only afterwards we add the corresponding write to the log. 

* Yet you say "writes gets logged in front of the reads" -- what am I missing?
* You also say "this order is different from what static analysis via 'getProgramEffects'" -- could you give an example?

Reply:
Sorry for the confusion, I have amended the `logEntry` function to append to the log instead of prepending.